api/v1alpha1/awschaos_webhook.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2020 Chaos Mesh Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// log is for logging in this package.
+var awschaoslog = logf.Log.WithName("awschaos-resource")
+
+// +kubebuilder:webhook:path=/mutate-chaos-mesh-org-v1alpha1-awschaos,mutating=true,failurePolicy=fail,groups=chaos-mesh.org,resources=awschaos,verbs=create;update,versions=v1alpha1,name=mawschaos.kb.io
+
+var _ webhook.Defaulter = &AwsChaos{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (in *AwsChaos) Default() {
+	awschaoslog.Info("default", "name", in.Name)
+}
+
+// +kubebuilder:webhook:verbs=create;update,path=/validate-chaos-mesh-org-v1alpha1-awschaos,mutating=false,failurePolicy=fail,groups=chaos-mesh.org,resources=awschaos,versions=v1alpha1,name=vawschaos.kb.io
+
+var _ ChaosValidator = &AwsChaos{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (in *AwsChaos) ValidateCreate() error {
+	awschaoslog.Info("validate create", "name", in.Name)
+	return in.Validate()
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (in *AwsChaos) ValidateUpdate(old runtime.Object) error {
+	awschaoslog.Info("validate update", "name", in.Name)
+	return in.Validate()
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (in *AwsChaos) ValidateDelete() error {
+	awschaoslog.Info("validate delete", "name", in.Name)
+
+	// Nothing to do?
+	return nil
+}
+
+// Validate validates chaos object
+func (in *AwsChaos) Validate() error {
+	specField := field.NewPath("spec")
+	allErrs := in.ValidateScheduler(specField)
+	allErrs = append(allErrs, in.ValidatePodMode(specField)...)
+	allErrs = append(allErrs, in.Spec.validateEbsVolume(specField.Child("volumeID"))...)
+	allErrs = append(allErrs, in.Spec.validateDeviceName(specField.Child("deviceName"))...)
```

Comment:
would be better add a validate instance when action is ec2-stop

Reply:
What is the purpose of this?

Reply:
if the action is ec2-stop, the instance field can be empty?

Reply:
No, this field should be required.